### ISSUE TITLE
Optimize the FFI backend

### ISSUE THREAD

**Issue by eregon** [ruby/prism#2402]
Currently it's slower.
I parse Prism's `lib/**/*.rb` files as a quick benchmark.
```
$ ruby -Ilib -rprism -rbenchmark -e '10.times { p Benchmark.realtime { Dir.glob("lib/**/*.rb") { |f| Prism.parse_file(f) } } }'
0.1726086970011238
0.1321143770001072
0.13798127100017155
0.147119974000816
0.14450125900111743
0.13631621999957133
0.128785977998632
0.15200302700031898
0.14583392900021863
0.14046238399896538

$ ruby --yjit -Ilib -rprism -rbenchmark -e '10.times { p Benchmark.realtime { Dir.glob("lib/**/*.rb") { |f| Prism.parse_file(f) } } }'
0.16340846000093734
0.11280311600057757
0.11687472400080878
0.11827147000076366
0.1305678140015516
0.11364481000055093
0.10130149700125912
0.10992031799833057
0.11991112200121279
0.13184594300037134

$ PRISM_FFI_BACKEND=true ruby -Ilib -rprism -rbenchmark -e '10.times { p Benchmark.realtime { Dir.glob("lib/**/*.rb") { |f| Prism.parse_file(f) } } }'
0.3652132739989611
0.3220108080022328
0.3004020099979243
0.3233778670000902
0.30800675899808994
0.31912518099852605
0.3041645149969554
0.29707527800201206
0.33191245600028196
0.29412771300121676

$ PRISM_FFI_BACKEND=true ruby --yjit -Ilib -rprism -rbenchmark -e '10.times { p Benchmark.realtime { Dir.glob("lib/**/*.rb") { |f| Prism.parse_file(f) } } }'
0.2793479979991389
0.21698390300298342
0.18590097600099398
0.20671838999987813
0.1923178380020545
0.1890159450013016
0.2051944070008176
0.18629215799956
0.1725655469999765
0.21039302300050622
```

```
$ PRISM_FFI_BACKEND=true ruby --yjit -Ilib -rprism -rbenchmark -rstackprof/autorun -e '10.times { p Benchmark.realtime { Dir.glob("lib/**/*.rb") { |f| Prism.parse_file(f) } } }'
0.27546833399901516
0.2049026849999791
0.1797586089996912
0.18482352300270577
0.19821838599818875
0.1960418029993889
0.18537640199792804
0.2127407010011666
0.18586592199790175
0.17573103299946524
StackProf results dumped at: /tmp/stackprof20240213-164464-1scane.dump

$ stackprof /tmp/stackprof20240213-164464-1scane.dump
==================================
  Mode: wall(1000)
  Samples: 1831 (8.40% miss rate)
  GC: 354 (19.33%)
==================================
     TOTAL    (pct)     SAMPLES    (pct)     FRAME
       476  (26.0%)         334  (18.2%)     Class#new
      1210  (66.1%)         234  (12.8%)     Prism::Serialize::Loader#load_node
       226  (12.3%)         226  (12.3%)     (marking)
       147   (8.0%)         147   (8.0%)     StringIO#getbyte
       252  (13.8%)         139   (7.6%)     Prism::Serialize::Loader#load_varuint
       128   (7.0%)         128   (7.0%)     (sweeping)
       114   (6.2%)         114   (6.2%)     Prism::LibRubyParser.pm_serialize_parse
        82   (4.5%)          82   (4.5%)     Prism::Location#initialize
      1311  (71.6%)          81   (4.4%)     Array#initialize
       576  (31.5%)          65   (3.5%)     Prism::Serialize::Loader#load_location
      1311  (71.6%)          35   (1.9%)     Array.new
      1205  (65.8%)          28   (1.5%)     Prism::Serialize::Loader#load_optional_node
        21   (1.1%)          21   (1.1%)     FFI::AbstractMemory#get_bytes
       246  (13.4%)          20   (1.1%)     Prism::Serialize::Loader#load_optional_location
        15   (0.8%)          15   (0.8%)     String#to_sym
        11   (0.6%)          11   (0.6%)     Prism::CallNode#initialize
        58   (3.2%)           9   (0.5%)     Prism::Serialize::Loader#load_comments
         8   (0.4%)           8   (0.4%)     Prism::DefNode#initialize
        44   (2.4%)           8   (0.4%)     Prism::Serialize::Loader#load_required_constant
         7   (0.4%)           7   (0.4%)     Integer#<<
        23   (1.3%)           7   (0.4%)     Prism::Serialize::Loader#load_string
         6   (0.3%)           6   (0.3%)     StringIO#pos=
        43   (2.3%)           6   (0.3%)     Prism::Serialize::Loader#load_line_offsets
         6   (0.3%)           6   (0.3%)     StringIO#read
         6   (0.3%)           6   (0.3%)     String#unpack1
        26   (1.4%)           6   (0.3%)     Prism::Serialize::Loader#load_constant
         5   (0.3%)           5   (0.3%)     Prism::ArgumentsNode#initialize
         5   (0.3%)           5   (0.3%)     Prism::RequiredParameterNode#initialize
         5   (0.3%)           5   (0.3%)     String#byteslice
         5   (0.3%)           5   (0.3%)     Kernel#require
```

Although of course the performance on CRuby of the FFI backend does not matter so much since there the C extension is used (by default). But it still matters for the `Prism.load` API.
So I'll get some numbers on TruffleRuby.
Related: https://github.com/oracle/truffleruby/issues/3293

**Comment by eregon** [comment]
On `truffleruby 23.1.2, like ruby 3.2.2, Oracle GraalVM Native [x86_64-linux]`, with https://github.com/ruby/prism/pull/2381 + `Array.new`:
```
$ PRISM_FFI_BACKEND=true ruby -Ilib -rprism -rbenchmark -e '30.times { p Benchmark.realtime { Dir.glob("lib/**/*.rb") { |f| Prism.parse_file(f) } } }'
0.630433598998934
0.2998583080006938
0.2663452880005934
0.22159439000097336
0.21022168499985128
0.2198489709990099
0.19900395200238563
0.20565986899964628
0.19073825899977237
0.20582825399833382
0.19916660799935926
0.18467618499926175
0.20206034500006353
0.18873625000196625
0.17961449599897605
0.19220407200191403
0.17271323099703295
0.1958472360020096
0.17021187099817325
0.1930403070000466
0.17139884399875882
0.17045755700019072
0.19395910000093863
0.1727715380002337
0.16954045799866435
0.18550562800010084
0.16578957800084027
0.16637309699945035
0.17212953499983996
0.1676743730022281
```

https://github.com/ruby/prism/pull/2381 helps a ton, it's around 2.7s per iteration before.

**Comment by eregon** [comment]
With https://github.com/ruby/prism/pull/2426, on a different machine:
```
$ ruby -v -Ilib -rprism -rbenchmark -e '10.times { p Benchmark.realtime { Dir.glob("lib/**/*.rb") { |f| Prism.parse_file(f) } } }' 
ruby 3.3.0 (2023-12-25 revision 5124f9ac75) [x86_64-linux]
0.14763125399986166
0.12022325599900796
0.10421211599896196
0.12528965499950573
0.10667782500240719
0.12413513700084877
0.12249747199894045
0.11835937800060492
0.10443235999991884
0.10476002499854076

$ ruby -v --yjit -Ilib -rprism -rbenchmark -e '10.times { p Benchmark.realtime { Dir.glob("lib/**/*.rb") { |f| Prism.parse_file(f) } } }'
ruby 3.3.0 (2023-12-25 revision 5124f9ac75) +YJIT [x86_64-linux]
0.14347519600050873
0.10289847100284533
0.08751282999946852
0.10573569500047597
0.08755050100080553
0.10640318600053433
0.10468406899963156
0.10008469699823763
0.08643773299991153
0.10499052399973152

$ ruby -v -Ilib -rprism -rbenchmark -e '100.times { p Benchmark.realtime { Dir.glob("lib/**/*.rb") { |f| Prism.parse_file(f) } } }'
truffleruby 24.1.0-dev-0e0df31f, like ruby 3.2.2, Oracle GraalVM JVM [x86_64-linux]
...
0.1016471110015118
0.12393867999708164
0.09867463500268059
0.09953019800013863
0.10812528300084523
0.10998337199998787
0.10899003600206925
0.13085450799917453
0.0991063209985441
0.0992955340007029

$ ruby -v -Ilib -rprism -rbenchmark -e '100.times { p Benchmark.realtime { Dir.glob("lib/**/*.rb") { |f| Prism.parse_file(f) } } }'
jruby 9.4.5.0 (3.1.4) 2023-11-02 1abae2700f OpenJDK 64-Bit Server VM 17.0.8+7 on 17.0.8+7 +jit [x86_64-linux]
...
0.17809201300042332
0.17167247199904523
0.16919517300266307
0.1693248859992309
0.16539225400265423
0.1730336239997996
0.16689628699896275
0.1696041690011043
0.16774694999912754
0.16836565999983577

$ ruby -Xcompile.invokedynamic=true -v -Ilib -rprism -rbenchmark -e '100.times { p Benchmark.realtime { Dir.glob("lib/**/*.rb") { |f| Prism.parse_file(f) } } }'
jruby 9.4.5.0 (3.1.4) 2023-11-02 1abae2700f OpenJDK 64-Bit Server VM 17.0.8+7 on 17.0.8+7 +indy +jit [x86_64-linux]
...
0.07872123400011333
0.0794329860000289
0.08794815900182584
0.08195378500022343
0.08005977499851724
0.0783269380008278
0.07669402199826436
0.07847700099955546
0.0762808360013878
0.07818350599700352
```

So it's as fast on TruffleRuby and JRuby+indy, using the FFI backend (which means extra serialization in C & deserialization in Ruby), as CRuby using the C extension.
So I think we can close this.

**Comment by eregon** [comment]
I'm reopening this to track doing something like #2428 when deserializing, i.e. avoiding to create Location objects eagerly on deserialization.

**Comment by kddnewton** [comment]
@eregon is this good to close now?

**Comment by eregon** [comment]
I plan to do another PR after #2531, I'll close myself when done

**Comment by kddnewton** [comment]
Sounds good

**Comment by eregon** [comment]
This is good now, simplified timings from https://github.com/ruby/prism/pull/2538#issue-2164156199:
```
$ PRISM_FFI_BACKEND=true ruby -v -Ilib -rprism -rbenchmark -e '300.times { p Benchmark.realtime { Dir.glob("lib/**/*.rb") { |f| Prism.parse_file(f) } } }'
  ruby 3.3.0:         0.251 (cext: 0.062)
  ruby 3.3.0 YJIT:    0.113 (cext: 0.053)
  truffleruby JVM:    0.054
  jruby 9.4.6.0:      0.219
  jruby 9.4.6.0 indy: 0.086
```
